elastic-agent-client.proto

```diff
@@ -101,26 +129,311 @@ message ActionResponse {
+    string token = 1;
+    // Transaction ID.
+    string txn_id = 2;
+    // Name of the key.
+    string name = 3;
+    // Value of the key.
+    bytes value = 4;
+    // Metadata of the key.
+    bytes metadata = 5;
+    // TTL of the key (in milliseconds)
+    uint64 ttl = 6;
+}
+
+// Response from `SetKey`.
+message StoreSetKeyResponse {
+    // Empty at the moment, defined for possibility of adding future return values.
+}
+
+// Deletes a key in the store.
+//
+// `txn_id` must be an ID of a transaction that was started with `READ_WRITE`.
+//
+// Does not error in the case that a key does not exist.
+message StoreDeleteKeyRequest {
+    // Token that is used to uniquely identify the connection to the Elastic Agent.
+    string token = 1;
+    // Transaction ID.
+    string txn_id = 2;
+    // Name of the key.
+    string name = 3;
+}
+
+// Response from `DeleteKey`.
+message StoreDeleteKeyResponse {
+    // Empty at the moment, defined for possibility of adding future return values.
+}
+
+// Commits the transaction in the store.
+//
+// Upon error the whole transaction is discarded so no need to call discard after error.
+message StoreCommitRequest {
+    // Token that is used to uniquely identify the connection to the Elastic Agent.
+    string token = 1;
+    // Transaction ID.
+    string txn_id = 2;
+}
+
+// Response from `CommitTxn`.
+message StoreCommitTxnResponse {
+    // Empty at the moment, defined for possibility of adding future return values.
+}
+
+// Discards the transaction in the store.
+message StoreDiscardRequest {
+    // Token that is used to uniquely identify the connection to the Elastic Agent.
+    string token = 1;
+    // Transaction ID.
+    string txn_id = 2;
+}
+
+// Response from `DiscardTxn`.
+message StoreDiscardTxnResponse {
+    // Empty at the moment, defined for possibility of adding future return values.
+}
+
+service ElasticAgentStore {
+    // Key-Value state storage is provided for each unit.
+    //
+    // Transactional store is provided to allow multiple key operations to occur before a commit to ensure consistent
+    // state when multiple keys make up the state of an units persistent state.
+    rpc BeginTxn(StoreBeginTxnRequest) returns (StoreBeginTxnResponse);
+    rpc GetKey(StoreGetKeyRequest) returns (StoreGetKeyResponse);
+    rpc SetKey(StoreSetKeyRequest) returns (StoreSetKeyResponse);
+    rpc DeleteKey(StoreDeleteKeyRequest) returns (StoreDeleteKeyResponse);
+    rpc CommitTxn(StoreCommitRequest) returns (StoreCommitTxnResponse);
+    rpc DiscardTxn(StoreDiscardRequest) returns (StoreDiscardTxnResponse);
+}
+
+// Requests an artifact from the Elastic Agent.
+message ArtifactFetchRequest {
+    // Token that is used to uniquely identify the collection of inputs to the agent. When started this is provided
+    // in the `ConnInfo`.
+    string token = 1;
+    // ID of the artifact.
+    string id = 2;
+    // SHA256 of the artifact.
+    string sha256 = 3;
```

Comment:
How would the client already know the SHA256 of the artifact? During the build we'd embed all valid artifact IDs and SHAs?

Reply:
This is how the protocol works. Best to look up how the Fleet Server artifact API works, don't want to conflate this PR with that information.